src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -2489,6 +2509,22 @@ bool QgsOgrProvider::changeGeometryValues( const QgsGeometryMap &geometry_map )
 
   const bool inTransaction = startTransaction();
 
+  // Some drivers may need to call ResetReading() after GetFeature(), such
+  // as GPKG in GDAL < 2.3.0 to avoid letting the database in a locked state.
```

Comment:
Maybe move this duplicate code out to a function?

Reply:
yeah, waiting for @elpaso confirmation that it actually helps for the issue we discussed by email

Reply:
Thanks, I'll try that tomorrow and it will certainly help but I'm 100% sure that's not enough: QgsOgrFeatureIterator constructor is calling reset in several places through OGRGeoPackageTableLayer::SetSpatialFilter, OGRGeoPackageTableLayer::SetSpatialFilterRect and OGRGeoPackageTableLayer::SetAttributeFilter they all call ResetReading and these methods are invoked in the QgsOgrFeatureIterator constructor.

Reply:
... and the initial rewind call of course.

Reply:
Should we disable geopackage transaction support until this is sorted out?

Reply:
@nyalldawson yes, if we can't find a solution before the release.

Reply:
@rouault I can confirm that this patch does **not** fix the issue. The `QgsOgrFeatureIterator` ctor needs patches too, I'll try to work on it.

Reply:
@elpaso For my information, could you explain or reexplain me what is the issue with QgsOgrFeatureIterator issuing ResetReading() at the beginning of the iteration ? Of course, it is problematic if ResetReading() is called after iteration has started, but not at its beginning. I probably miss something.

In the scenario you pointed in https://github.com/qgis/QGIS/issues/39178#issuecomment-705456951, if changeAttributeValue() no longer issues a ResetReading(), we should be safe, shouldn't we ?
```
        self.assertTrue(vl.startEditing())

        for f in vl.getFeatures():
            self.assertTrue(vl.changeAttributeValue(1, 1, 'new value'))
```

Reply:
@rouault no, the fact is that the last line of code here creates a new (nested) iterator
```c++
QgsVectorLayerUndoPassthroughCommandChangeAttribute::QgsVectorLayerUndoPassthroughCommandChangeAttribute( QgsVectorLayerEditBuffer *buffer, QgsFeatureId fid, int field, const QVariant &newValue )
  : QgsVectorLayerUndoPassthroughCommand( buffer, QObject::tr( "change attribute value" ) )
  , mFid( fid )
  , mField( field )
  , mNewValue( newValue )
  , mOldValue( mBuffer->L->getFeature( mFid ).attribute( field ) )
```

A commit worth more than a thousand words, this fixes the issue in the test:  https://github.com/qgis/QGIS/commit/505ef99328f8d1ba63919aeb89d815a6c4ce679b

Reply:
@rouault this PR fixes the issue https://github.com/qgis/QGIS/pull/39257 but it's just an ugly patch, I know it.

Reply:
> no, the fact is that the last line of code here creates a new (nested) iterator

That was the part I was missing, thanks.